rfc#0001

```diff
@@ -0,0 +1,140 @@
+packages into a prepared Keep. 
+
+## Motivation
+
+Once a Keep is prepared with microkernel, WASM runtime, etc., the
+workload (WASM binary + data) needs to be loaded into it.  A
+description is required.
+
+## Tutorial
+
+A Keep is the fundamental runtime component of Enarx: a TEE instance
++ Enarx runtime pieces, including WebAssembly and WASI layers.  In
+order for it to execute a workload, the WASM (WebAssembly) package
+that comprises it (binary/binaries + data elements) need to be loaded
+into it.  This is sent to it by the Enarx client agent (ECA),
+encrypted under a session key established at Keep provisioning time,
+and communicated to the ECA - the workload provisioning encryption
+key (WPEK).  The Keep has an agent running which is listening on a
+well-known port, or on a port number previously established during
+the Keep preparation phase.  The ECA connects to this port, and
+establishes a connection to the Keep.  The Keep takes the workload,
+decrypts it, and runs it.
+
+For this design to work, a route between the ECA and the Keep control
+plane needs to be available.  Note that this does not necessarily
+need to be direct: since an encrypted connection is not explicitly
+required (see below), proxying is possible.  This could, if required,
+be via an (untrusted) service on the host.
+
+One of the Enarx project's design principles is that all network
+and storage connections in and out of a Keep are encrypted.  In the
+case of network connections, this usually means that the session is
+encrypted using TLS (or dTLS), but since we may not have established
+any certificates in the Keep at this point, it cannot act as the
+server.  However, as the workload provisioning encryption key (WPEK)
+has already been communicated to the Keep, if the workload can be
+decrypted using this key, this assures the Keep that it is
+communicating with the ECA, or at least an entity in the same trust
+domain, so identity has been confirmed.
+
+There are possible use cases where a Keep might wish to cache a copy
+of the workload, either for reloading ("recycling" a Keep) or to
+transfer to another Keep in the same trust domain.  This caching, if
+under the origincal WPEK, is at the expense of Perfect Forward
+Secrecy.  If a Keep is to transfer the workload to another Keep in
+the same domain, it MAY encrypt it under a new WPEK.  If the
+original is used, it will need to established in the new Keep either
+at provisioning time (e.g. by the ECA) or at a later date through
+key negotiation.  Such functionality is considered risky, as the key
+management required is difficult to get right, and provides new
+options for compromise or design/implementation errors.  Any design
+of such functionality requires a new RFC (see "Future possibilities"
+below).
+
+The Keep MUST NOT accept unencrypted workloads, as there is no way
+to verify their provenance.
+The Keep SHOULD listen on a well-known port (see "Unresolved
```

Comment:
This involves the host to bind this well-known port and forward all incoming packets to the Keep in question. I'm not sure if I suggest the wording here to include this fact.

Reply:
That depends rather on the networking set-up that we're assuming.  I'm guessing by this point that the Keep already has an IP address and route to it, so binding isn't an issue hopefully.

Reply:
Instead of speaking about ports which sounds like implementation language to me, could we use a more general term, e.g. communication endpoint?

Reply:
I think we're specifically talking network ports for this design.  We can certainly discuss, however.  And that doesn't mean that our initial implementation will look like that.

Reply:
Would like @npmccallum to weigh in here.  There's a balance between over-vagueness and enough for us to move forward with a good implementation.  We've always assumed network connections, so unless there's a good reason to move away from this, that's my preference.